### ISSUE TITLE
[BUG] html report cannot by used by Jenkins' HTML Publisher

### ISSUE THREAD

**Issue by TimeInvestor** [microsoft/playwright#10274]
The html report generated by Playwright cannot by used by Jenkins' HTML Publisher

![image](https://user-images.githubusercontent.com/4818798/141434836-d61107f3-8557-4e8a-83a8-4f6eb5a2ce73.png)
![image](https://user-images.githubusercontent.com/4818798/141434928-5d4506d6-1bda-45b2-a885-3173f38ab9e2.png)

**Comment by DavertMik** [comment]
This is because Playwright does not generate raw HTML output, it creates a single page application with data stored in json and starts a web server to display this. At least this is what I see now 

![image](https://user-images.githubusercontent.com/220264/141663268-df42486b-96ab-4c01-a24d-aab47ca6de7e.png)

If I open the HTML itself in prints this information:

![image](https://user-images.githubusercontent.com/220264/141663285-9cf2fe30-7fec-40ee-8869-ce13423a581c.png)

As far as I remember Jenkins HTML publisher can show only raw HTML with no extra dependencies.

However, you could write your own HTML reporter using Playwright API https://playwright.dev/docs/test-reporters#custom-reporters

**Comment by TimeInvestor** [comment]
Thanks @DavertMik for the info.

Hope Playwright could produce raw HTML report as this kind of usage is common for CI platforms.

**Comment by Mischi** [comment]
For me, the `HTML Publisher plugin Version 1.25` works prefectly fine.

```
publishHTML([
    allowMissing: false,
    alwaysLinkToLastBuild: true,
    keepAll: true,
    reportDir: 'playwright-report',
    reportFiles: 'index.html',
    reportName: 'Playwright Report',
    reportTitles: ''
])
```

**Comment by TimeInvestor** [comment]
> For me, the `HTML Publisher plugin Version 1.25` works prefectly fine.
> 
> ```
> publishHTML([
>     allowMissing: false,
>     alwaysLinkToLastBuild: true,
>     keepAll: true,
>     reportDir: 'playwright-report',
>     reportFiles: 'index.html',
>     reportName: 'Playwright Report',
>     reportTitles: ''
> ])
> ```

@Mischi I am using the latest HTML publisher plugin for my Jenkins: version 1.28 (https://plugins.jenkins.io/htmlpublisher/). I use it as post-build step as a normal job, not in Jenkins pipeline.

BTW, what's your Playwright version?

**Comment by john24144** [comment]
I had a similar issue until today. For me, I found out I was getting a console error when the HTML publisher plugin was trying to display the report. It was the same error as this one mentioned here: https://stackoverflow.com/questions/34315723/jenkins-error-blocked-script-execution-in-url-because-the-documents-frame

**Comment by TimeInvestor** [comment]
> I had a similar issue until today. For me, I found out I was getting a console error when the HTML publisher plugin was trying to display the report. It was the same error as this one mentioned here: https://stackoverflow.com/questions/34315723/jenkins-error-blocked-script-execution-in-url-because-the-documents-frame

Thanks @john24144 
After setting `System.setProperty("hudson.model.DirectoryBrowserSupport.CSP", "")` via Jenkins Admin Script Console according to the stackoverflow answers, the Jenkins job could display Playwright's html report 😄 

I will close this issue for now.

**Comment by TimeInvestor** [comment]
> @TimeInvestor Removing the CSP header completely is a very unsafe operation and is not a secure way of dealing with this problem. This should be reopened. Reporters like mochawesome and Allure build the report with scripts and css assets linked through supporting files, not inline. Thus, they do not violate Jenkins default CSP and display properly via the publishHTML plugin.

So do you mean Playwright's html reporter should do the same?

**Comment by crazycabo** [comment]
Yes, simply linking required assets like scripts and such should elevate this issue in Jenkins and other platforms invoking CSP’s.

**Comment by crazycabo** [comment]
> > @TimeInvestor Removing the CSP header completely is a very unsafe operation and is not a secure way of dealing with this problem. This should be reopened. Reporters like mochawesome and Allure build the report with scripts and css assets linked through supporting files, not inline. Thus, they do not violate Jenkins default CSP and display properly via the publishHTML plugin.
> 
> So do you mean Playwright's html reporter should do the same?

Yes, simply linking required assets like scripts and such should elevate this issue in Jenkins and other platforms invoking CSP’s.

**Comment by TimeInvestor** [comment]
Reopen the issue as @crazycabo recommend that we should improve Playwright's html report to properly fix the issue instead of playing with Jenkins CSP.

**Comment by mstepin** [comment]
The security policy script in Jenkins _DOES_ fix it, but as mentioned above it isn't exactly the safest way to handle it. It is still better address within the HTML itself. I'd suggest reopening this issue pls.

**Comment by usmonster** [comment]
The latest release of Jenkins contains a fix for what seems like [a related issue](https://issues.jenkins.io/browse/JENKINS-71366)... Is anyone able to test whether it also fixes this issue? Thanks!

**Comment by coding-finger** [comment]
> The latest release of Jenkins contains a fix for what seems like [a related issue](https://issues.jenkins.io/browse/JENKINS-71366)... Is anyone able to test whether it also fixes this issue? Thanks!

This will not be fixed with the default CSP settings of Jenkins. They have just added the "allow-same-origin" attribute, as you can see in the commit: https://github.com/jenkinsci/jenkins/pull/8273/commits/5aed31687a12fda8ca219b5445f5b63e6758f849

After adding this to Jenkins, other problems will pop up:
`System.setProperty("hudson.model.DirectoryBrowserSupport.CSP", "sandbox allow-same-origin; default-src 'none'; img-src 'self'; style-src 'self';")`

**next ERROR:**
`Blocked script execution in '....' because the document's frame is sandboxed and the 'allow-scripts' permission is not set.`

So, after adding the allow-scripts, some more problems occur:
`System.setProperty("hudson.model.DirectoryBrowserSupport.CSP", "sandbox allow-same-origin allow-scripts; default-src 'none'; img-src 'self'; style-src 'self';")`

**next ERROR:**
`Refused to execute inline script because it violates the following Content Security Policy directive: "default-src 'none'". Either the 'unsafe-inline' keyword, a hash ('sha256-xB2htkE4msXNmN+SyWT+TbKZnio7AmRXBAVbKB1YTp8='), or a nonce ('nonce-...') is required to enable inline execution. Note also that 'script-src' was not explicitly set, so 'default-src' is used as a fallback.`

So, I tried to fix them:
`System.setProperty("hudson.model.DirectoryBrowserSupport.CSP", "sandbox allow-same-origin allow-scripts; default-src 'none'; img-src 'self'; style-src 'self';")`

**next ERROR:**
`Refused to apply inline style because it violates the following Content Security Policy directive: "style-src 'self'". Either the 'unsafe-inline' keyword, a hash ('sha256-TkjV6jmdsnq2swbyIbTx/+dY+rC1xle5YOhj7Pxl4lc='), or a nonce ('nonce-...') is required to enable inline execution.`

That worked, but after navigating to an error with a video, the next error popped up:
`System.setProperty("hudson.model.DirectoryBrowserSupport.CSP", "sandbox allow-same-origin allow-scripts; default-src 'none'; img-src 'self'; style-src 'self' 'unsafe-inline'; script-src 'unsafe-inline';")`

**next ERROR:**
`Refused to load media from '....someLink.webm' because it violates the following Content Security Policy directive: "default-src 'none'". Note that 'media-src' was not explicitly set, so 'default-src' is used as a fallback.`

Next, there was a problem with an image:
`System.setProperty("hudson.model.DirectoryBrowserSupport.CSP", "sandbox allow-same-origin allow-scripts; default-src 'none'; img-src 'self'; style-src 'self' 'unsafe-inline'; script-src 'unsafe-inline'; media-src 'self';")`

**next ERROR:**
`Refused to load the image 'data:image/png;base64,asdfasdf5CYII=' because it violates the following Content Security Policy directive: "img-src 'self'".`

After fixing the image, the pop-up is the next problem:
`System.setProperty("hudson.model.DirectoryBrowserSupport.CSP", "sandbox allow-same-origin allow-scripts; default-src 'none'; img-src 'self' data:; style-src 'self' 'unsafe-inline'; script-src 'unsafe-inline'; media-src 'self'")`

**next ERROR:**
`/jenkins/job/SWFE/job/playwright-starter_CI/Playwright_20Test_20Report/index.html#?testId=asdf:1 Blocked opening 'asdf5be22fbd0d.zip' in a new window because the request was made in a sandboxed frame whose 'allow-popups' permission is not set.`

After fixing the pop-up and trying to open the trace:
`System.setProperty("hudson.model.DirectoryBrowserSupport.CSP", "sandbox allow-same-origin allow-scripts allow-popups; default-src 'none'; img-src 'self' data:; style-src 'self' 'unsafe-inline'; script-src 'unsafe-inline'; media-src 'self'")`

**next ERROR:**
`Refused to load the script 'asdft/trace/index.6c78d458.js' because it violates the following Content Security Policy directive: "script-src 'unsafe-inline'". Note that 'script-src-elem' was not explicitly set, so 'script-src' is used as a fallback.`

The next problem is with the font:
`System.setProperty("hudson.model.DirectoryBrowserSupport.CSP", "sandbox allow-same-origin allow-scripts allow-popups; default-src 'none'; img-src 'self' data:; style-src 'self' 'unsafe-inline'; script-src 'self' 'unsafe-inline'; media-src 'self'")`

**next ERROR:**
`Refused to load the font 'asdf/trace/codicon.dcd00fb4.ttf' because it violates the following Content Security Policy directive: "default-src 'none'". Note that 'font-src' was not explicitly set, so 'default-src' is used as a fallback.`

So I changed the default-src to 'self', but still, there are problems with the frame-src:
`System.setProperty("hudson.model.DirectoryBrowserSupport.CSP", "sandbox allow-same-origin allow-scripts allow-popups; default-src 'self'; img-src 'self' data:; style-src 'self' 'unsafe-inline'; script-src 'self' 'unsafe-inline'; media-src 'self'; font-src 'self';")`

**next ERROR:**
`Refused to frame '' because it violates the following Content Security Policy directive: "default-src 'self'". Note that 'frame-src' was not explicitly set, so 'default-src' is used as a fallback.`

`System.setProperty("hudson.model.DirectoryBrowserSupport.CSP", "sandbox allow-same-origin allow-scripts allow-popups; default-src 'self'; img-src 'self' data:; style-src 'self' 'unsafe-inline'; script-src 'self' 'unsafe-inline'; media-src 'self'; font-src 'self'; frame-src 'self' data:;")`

And all of this leads to the final configuration:
`System.setProperty("hudson.model.DirectoryBrowserSupport.CSP", "sandbox allow-same-origin allow-scripts allow-popups; default-src 'self'; img-src 'self' data:; style-src 'self' 'unsafe-inline'; script-src 'self' 'unsafe-inline'; media-src 'self'; font-src 'self'; frame-src 'self' data:;")`

However, as others have pointed out, this is not an ideal solution, but rather a temporary workaround. I also suggest reopening this issue and evaluating other solutions for this problem.

**Comment by kutzi** [comment]
Thanks @coding-finger that was really helpful. I realised howerver that downloading of the traces doesn't work (with Chrome)
allow-downloads was missing:
`System.setProperty("hudson.model.DirectoryBrowserSupport.CSP", "sandbox allow-same-origin allow-scripts allow-popups allow-downloads; default-src 'self'; img-src 'self' data:; style-src 'self' 'unsafe-inline'; script-src 'self' 'unsafe-inline'; media-src 'self'; font-src 'self'; frame-src 'self' data:;")`
